handsanitizer/LLVMExtractor.cpp

```diff
@@ -0,0 +1,363 @@
+                                               std::vector<handsanitizer::Argument> &args_of_func,
+                                               std::shared_ptr<DeclarationManager> declarationManager) {
+        Type* retType;
+        bool sret = false;
+        for(auto& arg : args_of_func)
+            if(arg.isSRet){
+                retType = arg.getType();
+                sret = true;
+            }
+        if(!sret)
+          retType = ConvertLLVMTypeToHandsanitizerType(declarationManager, f.getReturnType());
+        return retType;
+    }
+
+    Purity ModuleFromLLVMModuleFactory::getPurityOfFunction(const llvm::Function &f) {
+        Purity p;
+        if (f.doesNotAccessMemory())
+            p = Purity::READ_NONE;
+        else if (f.doesNotReadMemory())
+            p = Purity::WRITE_ONLY;
+        else
+            p = Purity::IMPURE;
+        return p;
+    }
+
+
+    void ModuleFromLLVMModuleFactory::DeclareGlobalsInManager(std::shared_ptr<DeclarationManager> declarationManager,
+                                                              std::unique_ptr<llvm::Module> const &llvm_mod) {
+        std::vector<handsanitizer::GlobalVariable> globals;
+        for (auto &global : llvm_mod->global_values()) {
+            if (shouldGlobalBeIncluded(global)) {
+                auto globalType = this->ConvertLLVMTypeToHandsanitizerType(declarationManager, global.getType()->getPointerElementType());
+                GlobalVariable globalVariable(global.getName().str(), globalType);
+                declarationManager->addDeclaration(globalVariable);
+            }
+        }
+    }
+
+    bool ModuleFromLLVMModuleFactory::shouldGlobalBeIncluded(const llvm::GlobalValue &global) const {
+        return !global.getType()->isFunctionTy() &&
+                       !(global.getType()->isPointerTy() && global.getType()->getPointerElementType()->isFunctionTy()) &&
+                       !global.isPrivateLinkage(global.getLinkage()) &&
+                       !global.getType()->isMetadataTy() &&
+                       global.isDSOLocal();
+    }
+
+    Function::Function(const std::string &name, Type *retType, std::vector<Argument> arguments, Purity purity)
+            : name(name), retType(retType), arguments(arguments), purity(purity) {}
+
+    std::string Function::getFunctionSignature() {
+        return this->retType->getCTypeName() + " " + this->name + "(" + this->getTypedArgumentNames() + ");";
+    }
+
+    std::string Function::getTypedArgumentNames() {
+        std::stringstream output;
+        for(auto& args : this->arguments){
+            output << args.getType()->getCTypeName() << " " << args.getName() << ", ";
+        }
+        auto retstring = output.str();
+        if(!retstring.empty()){
+            retstring.pop_back(); //remove last ,<space>
+            retstring.pop_back();
+        }
+        return retstring;
+    }
+
+    std::string Function::getPurityName() {
```

Comment:
Consider lifting the string literals here to a class-level constant map, and changing the signature:

```suggestion
    const std::string& Function::getPurityName() const {
```

(Otherwise, this produces a lot of unnecessary copies.)